### ISSUE TITLE
Use hooks to implement `@observer` and `<Observer>`

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx-react#588]
Hooks (will be introduced in react 16.7) form a super solid foundation to write function based components. 

An early PoC is here: https://codesandbox.io/s/k2vmjpqvnv

We could hooks support here, or expose it as separate clean package. In the latter case that package could be pretty lightweight and aimed at React 16.7 purely:

* No Provider / inject (use `useContext` instead, which is better typeable)
* No propTypes (use TypeScript)
* No class component support (use this package). Removes most of the householding code, which is aimed at class patching
* No properties that are automatically converted to an observable object (restores the problem that computed values shouldn't use props)
* Limit devTools support to obtaining dependency tree. Measuring renderings and showing that on screen is now perfectly supported in the standard React devtools.

**Comment by johot** [comment]
Can't wait for this! I'm the frontend lead at my company (a bank with > 200 employees), I have built and maintain our component library and we use MobX for state in all our customer facing web applications (and love it). I would really like to show the team React 16.7 hooks together with MobX!

Regarding one smaller package vs integrating into `mobx-react` I think we should do both!

Start with a standalone alpha package that we can use and test right away, something like `mobx-react-hooks` or `mobx-react-light`.

Then we can all help out testing it and start using it now. Then in the future `mobx-react` could simply use this library internally and we will have the best of both worlds!

If such a small package could be created and published as an alpha like *now* 😄 I would love helping out and testing it right away 😃

**Comment by mweststrate** [comment]
Hey! If you would be interested in forking this repo and removing /
converting tests that would already be a great start!

Op za 27 okt. 2018 09:37 schreef Johan Otterud <notifications@github.com>:

**Comment by johot** [comment]
So you are thinking forking `mobx-react` and making it into the smaller package you described? Stripping out stuff not needed as you described? Also taking the code from your code sandbox and inserting it?

**Comment by mweststrate** [comment]
Yeah I think that should work indeed

Op za 27 okt. 2018 om 10:16 schreef Johan Otterud <notifications@github.com

**Comment by johot** [comment]
Will give it a shot later tonight then 👍

**Comment by johot** [comment]
Ok! 

Now I have created a fork of `mobx-react` and named it `mobx-react-hooks` (for now anyways). It can be found here: <https://github.com/johot/mobx-react-hooks>. I basically did what @mweststrate said and stripped away some features like Provider, inject etc. I then copied the new version from the code sandbox.

I am not yet sure how much to strip away, there were something called EventEmitters and reporters that I kept, not sure what they do. For the record I have not worked in the mobxcode base before.

I also disabled the `index.d.ts` temporarily just to get something up and working. I also kept the old `observer.js` implementation in an `.old` file for easy reference reading.

So to try this out:
* clone repo <https://github.com/johot/mobx-react-hooks>
* `npm run build`
* Copy the files generated in the root folder to a `<your_project>\node_modules\mobx-react-hooks` directory (or does a simpler way exist without uploading to `npm`?).

* Now you can use it like this:

```jsx
import * as React from "react";
import { useObservable, observer } from "mobx-react-hooks";

export const Counter = observer(props => {
  const counter = useObservable({
    count: 0
  });
  return (
    <div>
      {counter.count}
      <button onClick={() => counter.count--}>-</button>
      <button onClick={() => counter.count++}>+</button>
      <hr />
    </div>
  );
});
```

Further directions @mweststrate? Also a bunch of the tests will not run right now, but I still need more time to learn about the code and what to keep/remove.

**Comment by mohamedmansour** [comment]
The code is so much cleaner!

**Comment by xaviergonz** [comment]
> No properties that are automatically converted to an observable object (restores the problem that computed values shouldn't use props)

shouldn't there be a hook to turn props into observable?

**Comment by johot** [comment]
Created a small repo if you want to play around with the prototype @mweststrate created. Needed it for a tech talk at work. Uses CRA 2.1 with TypeScript support + typings for React 16.7.0-alpha hooks. Thanks to @steadicat for the typings!

Repo:
https://github.com/johot/create-react-app-2-typescript-react-hooks-mobx

**Comment by FredyC** [comment]
@johot Thanks for the work you are doing. I run into some strange obstacle. If I use `observer` from your fork, I end up with this error. Switching back to `mobx-react` works so it must be one of your changes causing this.

![image](https://user-images.githubusercontent.com/1096340/47749253-0c51ff80-dc8d-11e8-9623-f2edeb6aafcc.png)

**Comment by johot** [comment]
@FredyC yes I also got that one today... not sure why I didn't get it earlier... I'll see if I have time to look into it, you could temporarily remove those lines to make it working (thats what I did today).

**Comment by nassimbenkirane** [comment]
@johot , @FredyC, in the original version, there's a declaration for isDevToolsEnabled 
I managed to have it working by adding those two lines : 
```javascript
let isDevToolsEnabled = false
let isUsingStaticRendering = false
```
Thanks for your repo :)

**Comment by johot** [comment]
The big question is still open, we basically have two options:

1. A `mobx-react-hooks` (or something similair) package that only works with React 16.7
2. Add hooks functionality to `mobx-react` for example as pointed out by @nassimbenkirane we would add this code into it's own export so you would do:
`import { observer, useObservable } from 'mobx-react/hooks'`. Not sure if this change to `mobx-react` would break if you don't use React 16.7?

I do see many (such as myself) having projects that depend on `mobx-react` and have many existing class components, using approach 2 it would be easier to migrate to hooks. A separate package could get a bit more confusing? On the other hand remembering that there is a difference between `observer` from `mobx-react` and `mobx-react/hooks` could also be confusing...

Have you considered this anymore @mweststrate?

**Comment by Gameghostify** [comment]
It would make sense to add to mobx-react instead of creating a new repo, similar to how `Provider` and `inject` were added later on.

And, if hooks get added to react core, it'd make sense for all react-related mobx functionality to be in a single package - `mobx-react`.

(However, the difference between `observer from "mobx-react` and `mobx-react/hooks`, as @johot noted,  could be really confusing 😕 )

**Comment by lennerd** [comment]
Dumb question maybe: but wouldn't it be possible to have one function/decorator for both? The incoming component is either a function or a class, no?

**Comment by lennerd** [comment]
> Once the hooks are finalized and released, we can do breaking release of mobx-react and at that point, it should be mostly about find & replace.

@FredyC What do you mean with breaking changes? Hooks will coexist with component classes (for a while at least, also to ease the transition and learning process). Why should `mobx-react` be any different supporting both concepts?

@johot Feature detection inside `mobx-react` might not be needed here, if we use proper version numbers. Or am I missing something?

Maybe I'm misunderstanding the discussion. If you want to remove the support for decorating classes with `observer` than this library is a lot less useful to projects using both concepts IMHO. Supporting both hooks and classes like React handles it will bring the same advantages to `mobx-react` like its bringing to the `react` library. Am I missing something?

**Comment by johot** [comment]
@lennerd I think you are correct, you mean we could simply say that `mobx-react 5.4` (or whatever the version is) would require React 16.7+?

The simplest most straightforward solution would be to add one more export:
`import { useObservable } from 'mobx-react'`

The question then becomes what implementation of the `observer` do we use? We have three options I believe:

1. Use the new `observer` implementation (implemented using a functional component with hooks).
2. Use the original class based implementation (will work with hook based functional components, I have tried it).
3. Default to the original class based one but can opt in to use the new one by doing `import { observer } from 'mobx-react/hooks'` or something. Could be a bit confusing.

Would currently lean against option 1. but not sure if it could break something?

**Comment by johot** [comment]
*Further investigation results:*
It seems we need to use the new `observer` implementation that is not implemented using a class, if not you will get the error message:

`Hooks can only be called inside the body of a function component.`

If trying to wrap a functional component with hooks using an observer. This means we would need to use the solution nr. 1 posted above.

I might create a new repo for you to test and see.

**Comment by lennerd** [comment]
Like I said it might be possible to use either the "hooked" or the "unhooked" version of the decorator depending on the passed component being a function or a class.

I really don't see why a big hype for a new way of working with React should mean this library needs two complete different implementation to work. I'm hyped about hooks as well, really! But I would love to see this library being able to work with both my functional and class components as seamlessly as possible to ease the process of switching to hooks, reusing old components of previous projects, etc.

Class components are not deprecated(, yet).

**Comment by johot** [comment]
> Like I said it might be possible to use either the "hooked" or the "unhooked" version of the decorator depending on the passed component being a function or a class.
> 
> I really don't see why a big hype for a new way of working with React should mean this library needs two complete different implementation to work. I'm hyped about hooks as well, really! But I would love to see this library being able to work with both my functional and class components as seamlessly as possible to ease the process of switching to hooks, reusing old components of previous projects, etc.
> 
> Class components are not deprecated(, yet).

I agree on everything you say!
And just like you say we could simply use the new implementation when a component is passed as a function and fall back to the old version for a class. In my current tests the code will fail if passing a class component to the new `observer` implementation so we definitely need something like that.

**Comment by FredyC** [comment]
@lennerd You are definitely right, that `mobx-react` shouldn't just throw away eg. `inject`, but it should be deprecated and removed in the next major. Otherwise, you would need to rewrite inject to use new Context and question is if its worth it.

**Comment by lennerd** [comment]
@FredyC Just took a peak at the code. Surprised it's still using the legacy context API. This makes the whole thing indeed more complex than I thought.

> Otherwise, you would need to rewrite inject to use new Context and question is if its worth it.

If class components will still be supported in React 17 I think there is no way around removing the legacy API anyway to support both component types. At least when using the same library for both.

**Comment by johot** [comment]
Ok everyone, I created a new fork of `mobx-react` which supports functional components with hooks + class components using the old implementation. Had to delete my old `mobx-react-hooks` repo since I can only fork one copy per github account, but it doesn't matter I think this is the correct way to go. Also added typings for `useObservable`.

Check out the fork here and see if you can get it running!
https://github.com/johot/mobx-react

**Comment by FredyC** [comment]
Looks good. I made some comments directly to your commit there. Would be nice if @mweststrate can publish it under `next` dist tag to make it easier to install.

**Comment by FredyC** [comment]
Looks good. I made some comments directly to your commit there. Would be nice if @mweststrate can publish it under `next` dist tag to make it easier to install.

Should we also deprecate old context? I mean this is totally possible to handle in-app and I am already doing it since `inject` is not needed so I can just have my own Provider. If the context would be added to the package, is there some real benefit to it?

**Comment by FredyC** [comment]
I am starting to think that we should deprecate Provider/inject completely and remove it in the next major which should support React 16.3+ only. Point is that there is no extra logic needed in terms of this package for a Provider as using React `createContext` directly is basically a one-liner.

The following can be just used in the userland and it's super easy and flexible. You can even have more of these which is great toward separation of concerns. Also, it allows much better type safety since it's possible to specify value type of the `Context` and then `useContext` just works with that specified type. That is currently a huge pain point with `inject` HOC.

```tsx
import { useContext, createContext } from 'react'

const Context = createContext<TStores>(createStores())

function App() {
  return <Context.Provider value={createStoresLater()}><MyComponent /></Context.Provider>
}

function MyComponent() {
  const stores = useContext(Context)
  // stores are already correctly typed without any extra effort
  return stores[name]
}
```

What do you think? Is there some logic you would expecting from a Provider?

**Comment by Albert-Gao** [comment]
> What do you think? Is there some logic you would expecting from a Provider?

Just tried it! Pretty good. Works like a charm. Interesting though, can only use the hook version `observer` function, if try to use the one from `mobx-react`, will have an error in the browser, `Hooks can only be called inside the body of a function component.`

For the updating part, my idea is there are 2 versions:
- one for react 16.7+. Which provides helper function using hooks api.
- one for current mobx-react

**Comment by FredyC** [comment]
Regarding observer, there is a fork from @johot ☝ which has a observer fixed for hooks, although at current version it wont work with regular stateless functions most likely. I am working on it to make it backward compatible.

**Comment by guojenman** [comment]
works great, the only issue i've found while playing w/ this is that because react-suspense is handle by throwing a Promise, mobx Reaction catches that as an error. For now i'm just swallowing the error with a 3rd argument when constructing the Reaction.

**Comment by FredyC** [comment]
I made a couple of small changes to @johot's fork, mostly about deprecating Provider/inject a mentioned above. Otherwise, there are no breaking changes. Unfortunately, some tests are failing while using a functional observer and I wasn't able to figure what's wrong. Would be lovely if someone can have a look.

I also want to share an alternate way for using forks which is also great for a team development. It's using [Yalc](https://github.com/whitecolor/yalc) tool. It should be installed globally.

1) clone the repo
2) install dependencies (preferably yarn)
3) yarn build
4) yalc publish (it's published locally)

In your project folder then just run `yalc add mobx-react`

Yalc actually makes the copy of a published build into your project folder, eg. `.yalc/mobx-react` and this folder is copied to node_modules and also added to `package.json`. If you are working on the team, you can just commit the `.yalc` folder to the repo and it will work for everyone without actually install Yalc or doing anything else.

**Comment by bdombro** [comment]
I am testing @johot's fork and am still getting "Hooks can only be called inside the body of a function component."

**Comment by FredyC** [comment]
@bdombro Well then and are sure you are using hooks inside function component body? :) It probably needs some example. I am using that fork in a medium app and it is working

**Comment by barbalex** [comment]
@bdombro there is a problem when using react-hot-loader that causes this error. I do not remember the link to the issue there. A solution is being worked on, but in the meantime the workaround is to add this to every file you use hooks in:

```js
import { setConfig } from 'react-hot-loader'
setConfig({ pureSFC: true })
```

**Comment by bdombro** [comment]
"...something about it occuring..." -> When you say 'it', do you mean a hook issue or a mobx-react->observer issue?

And curious, you're not using Webpack?

**Comment by barbalex** [comment]
Sorry for not being clear. I meant the `Hooks can only be called inside the body of a function component` issue when using hooks.

I am using above mentioned workaround in two gatsby apps. Gatsby uses webpack and react-hot-loader together.

**Comment by bdombro** [comment]
I added another branch to my [test app](https://github.com/bdombro/mobx-hook-test) to demonstrate that hooks are working.

There are now 3 branches:

* master -> observer test with hot reloading
* no-hot-reload -> observer test without hot reloading. This demonstrates that hot reload is not the cause of the issue.
* basic-hook-test -> same as no-hot-reload, except using state hooks instead of 
mobx or observer. This demonstrates that hooks are in fact working with the current build


@barbalex aside, are you sure that Gatsby uses Webpack? I cannot find any evidence of that and notice that the Gatsby CLI does not depend on Webpack --> https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-cli/package.json. It appears to use babel without webpack.

**Comment by barbalex** [comment]
@bdombro facepalm - you are right, I hadn't noticed. Did not know this type of functionality was possible without webpack

**Comment by bdombro** [comment]
@barbalex before you facepalm though, Gatsby does list Webpack as a feature on their homepage. Perhaps they do use it, but I didn't see it at first glance.

**Comment by FredyC** [comment]
@bdombro It works for me :) You are even missing the import of `useObservable` there so I am not sure how could you even get to the error with hooks really.

I really recommend [using Yalc](https://github.com/mobxjs/mobx-react/issues/588#issuecomment-435571367) tools for fiddling with forks. Doing it manually the way you have in readme is very volatile and you might not even realize that you not using correct version.

**Comment by FredyC** [comment]
I've zipped a working version with the fork. You still need to `yarn install`, but since the fork is referenced in `package.json` as `file:.yalc/mobx-react`, you can be sure that fork is really used. This is from the master branch with hot reloading.

[mobx-hook-test.zip](https://github.com/mobxjs/mobx-react/files/2546238/mobx-hook-test.zip)

**Comment by bdombro** [comment]
Thanks @FredyC ! I tried using yalc at first using your instructions but the instructions didn't quite work. Now you've convinced me to try to learn yalc harder. Very cool.

Btw, my app requires some files in the build directory to work, which was excluded in your zip. Below is your zip with the build dir added back.

[mobx-hook-test.zip](https://github.com/mobxjs/mobx-react/files/2546278/mobx-hook-test.zip)

**Comment by johot** [comment]
For those of you that want to test this easily, I created a release in the form of a `.tgz` file. This means it's very easy to install in your projects, just copy the file to your project root and do a `npm install mobx-react-5.4.0-alpha.tgz`

File can be found here:
https://github.com/johot/mobx-react/releases/tag/5.4.0-alpha

**Comment by mweststrate** [comment]
My proposal would be to release the hooks version as separate package, only React 16.7+ compatible. For example `mobx-react-lite`

1. In the current mobx-react package there is quite some legacy oriented stuff, and it would be a shame to build even more layers on top of it, making the package unneceseary heavy. 
2. The lite version could support function components only, which is I think the direction React is moving into in the long term
3. For code bases having class components, the classic package can still be used
4. Lite can be really lite: dropping Provider / Inject would not be a breaking change, and code bases could be slowly migrated over, until `mobx-react` is not needed anymore
5. I think having two automatic different implementations of `observer`, based on React version, is a really bad idea. It would mean that if someone upgrades from React 16.6 to 16.7 would cause mobx-react to behave very differently, without understanding why (that package was not upgraded after all). And the semantically differences might be very subtle, but still be there, because mobx-react suddenly starts to generate a a different kind of component (function instead of class), which might break unit tests, reflection, optimizations, references, static members, etc. etc. It will have unforeseen impact on consumers that do complicated things with consumers. #523 was reverted for a similar issues, and that was a smaller transition.

**Comment by FredyC** [comment]
Yea, considering I've spent like 2 hours trying to fix tests to make them work both ways, I am definitely for separating it :) It's also better for an slower adoption as you said, whenever you need to use hooks, you will import observer from a lite package.

Should we also still support `<Observer>`? I think that one is still handy when you need to watch for changes only in a part of the tree.

**Comment by mweststrate** [comment]
Yeah, I think it is still handy indeed, and probably only a few lines of
code, not requiring any dirty stuff :)

Op di 6 nov. 2018 om 13:12 schreef Daniel K. <notifications@github.com>:

**Comment by johot** [comment]
I can agree on this @mweststrate.

But it should still be possible to import both `mobx-react` and `mobx-react-lite` right? I know of some projects I am involved in where we wouldn't be able to convert all class component to functional components at once 😄

**Comment by guojenman** [comment]
**Current implementation breaks when used with Suspense**

Attached is minimal reproducible example

1. When you first run this, the component waits 3s before rendering (suspense)
2. now in App.js replace <Suspended/> with <ObservedAndSuspended />
3. [mobx] Encountered an uncaught exception that was thrown by a reaction or observer component, in: 'Reaction[observer(Suspended)] 
Promise {<pending>}



[test-mobx-hooks.zip](https://github.com/mobxjs/mobx-react/files/2553323/test-mobx-hooks.zip)

**Comment by mweststrate** [comment]
@johot yeah, that is the whole point, for legacy code bases you can use both (lite should be pretty small anyway), and for "future modern" non-class projects you could just use lite only

**Comment by mweststrate** [comment]
@johot, I've just invited you to the MobX organization. Would you mind renaming the project to `mobx-react-lite` and move it over, if you agree? That makes it easier to track issues independently, and enables any other mobx organization member to contribute directly to the repository.

**Comment by johot** [comment]
Yes ofc will do @mwestrate 👌

**Comment by johot** [comment]
@mweststrate Got the invite now thanks. However I can't transfer it since it's a fork of `mobx-react` which already exists in the organization. Should we start a clean repo instead or can I "unfork" it somehow?

**Comment by FredyC** [comment]
I think there is no need to keep history. Most of the things will go away anyway so I would go for a new repo directly under the `mobxjs`.

@mweststrate If you are going to do that, please give me access as well, I want to contribute to this too.

**Comment by johot** [comment]
@FredyC did some great work on the fork, definitely should be part of this 😄

**Comment by mweststrate** [comment]
@johot simplest way is probably to throw away `.git` dir and run `git init` again. 

@FredyC will send invite!
